sql: release table name as soon as it is no longer referenced

vivekmenezes opened the pull request:
When dropping a table, the table name would be released only after
the table was truncated. Table truncation can take a long time and
it doesn't make for a good user experience when a user expects a
name for a dropped table to be available almost immediately.

fixes #7348

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13908)
<!-- Reviewable:end -->

vivekmenezes committed:
sql: release table name as soon as it is no longer referenced

When dropping a table, the table name would be released only after
the table was truncated. Table truncation can take a long time and
it doesn't make for a good user experience when a user expects a
name for a dropped table to be available almost immediately.

fixes #7348

tamird commented:
Reviewed 3 of 3 files at r1.
Review status: 2 of 3 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/sql/drop.go, line 980 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13908#-KeZoQH4-cewlR9f1e_w:-KeZoQH4-cewlR9f1e_x:bwtujt8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3c72df67da7f237a0c8008be4b2cdbb3460a217/pkg/sql/drop.go#L980)):*
> ```Go
> 	if err := db.Txn(ctx, func(txn *client.Txn) error {
> 		b := &client.Batch{}
> 		// Use CPut because we want this function to be idempotent.
> ```

that's not what idempotent means; this function will fail on retry. this is to prevent trampling unexpected values. I'd just remove the comment.

---

*[pkg/sql/drop.go, line 983 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13908#-KeZpfmBNIP7EGo-OHaB:-KeZpfmBNIP7EGo-OHaC:b34zp5i) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3c72df67da7f237a0c8008be4b2cdbb3460a217/pkg/sql/drop.go#L983)):*
> ```Go
> 		b.CPut(nameKey, nil, tableDesc.ID)
> 		txn.SetSystemConfigTrigger()
> 		if err := txn.Run(b); err != nil {
> ```

```
err := txn.Run(b)
if _, ok := err.(*roachpb.ConditionFailedError); ok {
  return nil
}
return err
```

---

*[pkg/sql/drop.go, line 993 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13908#-KeZpyjUYd2Ch955MOsy:-KeZpyjVurPkoCcxHc1e:bwuxih0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3c72df67da7f237a0c8008be4b2cdbb3460a217/pkg/sql/drop.go#L993)):*
> ```Go
> 	}
> 
> 	if testingKnobs.RunAfterTableNameDropped != nil {
> ```

nit:

```
if cb := testingKnobs.RunAfterTableNameDropped; cb != nil {
  if err := cb(); err != nil {
    return err
  }
}
```

---

*[pkg/sql/drop_test.go, line 324 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13908#-KeZqkiGqAe2_rWP8PtG:-KeZqkiHzfRM5aImc6o4:b-xup2b6) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3c72df67da7f237a0c8008be4b2cdbb3460a217/pkg/sql/drop_test.go#L324)):*
> ```Go
> 				runAfterTableNameDropped = nil
> 				if runFunc != nil {
> 					go runFunc()
> ```

as suggested below, consider moving this `go` to the innermost part requiring asynchronous execution.

---

*[pkg/sql/drop_test.go, line 339 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13908#-KeZqH1OpafPyJZv8EbA:-KeZqH1OpafPyJZv8EbB:bigor1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3c72df67da7f237a0c8008be4b2cdbb3460a217/pkg/sql/drop_test.go#L339)):*
> ```Go
> 	numRows := 2*sql.TableTruncateChunkSize + 1
> 	createKVTable(t, sqlDB, numRows)
> 	var createWG sync.WaitGroup
> ```

this is an odd use for a waitgroup. how about an unbuffered error channel, instead? that also has the benefit of moving error handling to the main goroutine.

something like
```
runAfterTableNameDropped = func() {
_, err := sqlDB.Exec(`SELECT * FROM t.kv`)
go func() { errChan <- err }()
}
```
so as not to hold up the callback itself.

---

*[pkg/sql/drop_test.go, line 388 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13908#-KeZn3Xt_FgrAU1FGvse:-KeZn3Xt_FgrAU1FGvsf:b-98m2al) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3c72df67da7f237a0c8008be4b2cdbb3460a217/pkg/sql/drop_test.go#L388)):*
> ```Go
> 
> 	checkKeyCount(t, kvDB, tablePrefix, 3*numRows)
> 	var wg sync.WaitGroup
> ```

this is an odd use for a waitgroup. how about an unbuffered error channel, instead? that also has the benefit of moving error handling to the main goroutine.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13908)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
*[pkg/sql/drop.go, line 980 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13908#-KeZoQH4-cewlR9f1e_w:-Ke_NfiCmUAJi4O9K2J4:b-qitxh5) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3c72df67da7f237a0c8008be4b2cdbb3460a217/pkg/sql/drop.go#L980)):*
> err := txn.Run(b)
> if _, ok := err.(*roachpb.ConditionFailedError); ok {
>   return nil
> }
> return err

changed to

`
// Use CPut because we want to remove a specific name -> id map.
`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13908)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 2 of 3 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/sql/drop.go, line 983 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13908#-KeZpfmBNIP7EGo-OHaB:-Ke_Nsfh3lXTxK2p78Ks:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3c72df67da7f237a0c8008be4b2cdbb3460a217/pkg/sql/drop.go#L983)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

```
err := txn.Run(b)
if _, ok := err.(*roachpb.ConditionFailedError); ok {
  return nil
}
return err
```
</blockquote></details>

Done.

---

*[pkg/sql/drop.go, line 993 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13908#-KeZpyjUYd2Ch955MOsy:-Ke_O1bwqIo9-wKOimSj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3c72df67da7f237a0c8008be4b2cdbb3460a217/pkg/sql/drop.go#L993)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit:

```
if cb := testingKnobs.RunAfterTableNameDropped; cb != nil {
  if err := cb(); err != nil {
    return err
  }
}
```
</blockquote></details>

Done.

---

*[pkg/sql/drop_test.go, line 324 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13908#-KeZqkiGqAe2_rWP8PtG:-Ke_Y12m8qwyoN5Cme05:b1umkgl) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3c72df67da7f237a0c8008be4b2cdbb3460a217/pkg/sql/drop_test.go#L324)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

as suggested below, consider moving this `go` to the innermost part requiring asynchronous execution.
</blockquote></details>

Not sure what this suggestion is. I've changed the code to use a chan error.

---

*[pkg/sql/drop_test.go, line 339 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13908#-KeZqH1OpafPyJZv8EbA:-Ke_XvCOP3YUbxvImdrW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3c72df67da7f237a0c8008be4b2cdbb3460a217/pkg/sql/drop_test.go#L339)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this is an odd use for a waitgroup. how about an unbuffered error channel, instead? that also has the benefit of moving error handling to the main goroutine.

something like
```
runAfterTableNameDropped = func() {
_, err := sqlDB.Exec(`SELECT * FROM t.kv`)
go func() { errChan <- err }()
}
```
so as not to hold up the callback itself.
</blockquote></details>

Done.

---

*[pkg/sql/drop_test.go, line 388 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13908#-KeZn3Xt_FgrAU1FGvse:-Ke_XwTf-KFzQLPA-tdy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3c72df67da7f237a0c8008be4b2cdbb3460a217/pkg/sql/drop_test.go#L388)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this is an odd use for a waitgroup. how about an unbuffered error channel, instead? that also has the benefit of moving error handling to the main goroutine.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13908)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Does this also resolve #12123?

vivekmenezes commented:
It doesn't. We cannot resolve #12123 without some tradeoffs because the descriptors are cached across the cluster. In other words there is no transactional drop/create of a table, but we can change the behavior to be more user friendly.

andreimatei commented:
LGTM
Good idea with this change!

---

Review status: 1 of 3 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/sql/drop_test.go, line 398 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13908#-KeeL-QESZuboTMRmGs7:-KeeL-QESZuboTMRmGs8:b-bkitl6) ([raw file](https://github.com/cockroachdb/cockroach/blob/8781c42afa1cb9519209df0e535a7144aed94a88/pkg/sql/drop_test.go#L398)):*
> ```Go
> 	}
> 
> 	// Drop the newly created table.
> ```

maybe say in this comment that the hook recreated the table. Or even better, move this drop into the hook.

---

*[pkg/sql/drop_test.go, line 403 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13908#-KeeKHQIljMZhP1m1Fd7:-KeeKHQJTk0yPgwsPaeD:b-6spzl2) ([raw file](https://github.com/cockroachdb/cockroach/blob/8781c42afa1cb9519209df0e535a7144aed94a88/pkg/sql/drop_test.go#L403)):*
> ```Go
> 	}
> 
> 	// Test that deleted table cannot be used. This prevents regressions where
> ```

I think you can remove this assertion since you've copied it in the knob, right?

---

*[pkg/sql/drop_test.go, line 409 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13908#-KeeLFnsohMoVCPZG5eL:-KeeLFnsohMoVCPZG5eM:bswb2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/8781c42afa1cb9519209df0e535a7144aed94a88/pkg/sql/drop_test.go#L409)):*
> ```Go
> 	}
> 
> 	if gr, err := kvDB.Get(ctx, descKey); err != nil {
> ```

can this and the next checks be moved into the hook, before the table is recreated? If so, I  think that'd be better since the test would be a little tighter.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13908#-:-KeeJDuvCFyNCRxUZHyg:br988tu)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 1 of 3 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/sql/drop_test.go, line 398 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13908#-KeeL-QESZuboTMRmGs7:-KeeTNFtIco5L7eZI8xz:b-vnxa8d) ([raw file](https://github.com/cockroachdb/cockroach/blob/8781c42afa1cb9519209df0e535a7144aed94a88/pkg/sql/drop_test.go#L398)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

maybe say in this comment that the hook recreated the table. Or even better, move this drop into the hook.
</blockquote></details>

Moved the DROP completely. See comment below.

---

*[pkg/sql/drop_test.go, line 403 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13908#-KeeKHQIljMZhP1m1Fd7:-KeeThkiC1KAR66y-0gu:b-sjmtz5) ([raw file](https://github.com/cockroachdb/cockroach/blob/8781c42afa1cb9519209df0e535a7144aed94a88/pkg/sql/drop_test.go#L403)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think you can remove this assertion since you've copied it in the knob, right?
</blockquote></details>

True, and so now I do not need to even DROP the table any more. Removed Drop from the hook. The hook is only used to check that the name can be used.

---

*[pkg/sql/drop_test.go, line 409 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13908#-KeeLFnsohMoVCPZG5eL:-KeeVMkLI50wbq8pt_N9:b-10mc9) ([raw file](https://github.com/cockroachdb/cockroach/blob/8781c42afa1cb9519209df0e535a7144aed94a88/pkg/sql/drop_test.go#L409)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

can this and the next checks be moved into the hook, before the table is recreated? If so, I  think that'd be better since the test would be a little tighter.
</blockquote></details>

Only the nameKey can be moved there and I've done that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13908)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
I've also added a counter to check that the DROP table retry actually happens

---

Review status: 1 of 3 files reviewed at latest revision, 9 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13908#-:-KeeX686QDy8o64haimt:bt0gwoe)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/drop_test.go, line 385 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13908#-KeevAPp6hkbdbu67BxH:-KeevAPp6hkbdbu67BxI:b-e3wbce) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ae895e4217b2d919f539475751b5dd54cd6dbb1/pkg/sql/drop_test.go#L385)):*
> ```Go
> 			`SELECT * FROM t.kv`,
> 		); !testutils.IsError(err, `table "t.kv" does not exist`) {
> 			return errors.Wrap(err, "different error than expected")
> ```

this is incorrect. `errors.Wrap(nil, ...)` is `nil`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13908)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/drop_test.go, line 385 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/13908#-KeevAPp6hkbdbu67BxH:-Kef_Wl_VaVi2MLMdkwZ:b-q3aadx) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ae895e4217b2d919f539475751b5dd54cd6dbb1/pkg/sql/drop_test.go#L385)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this is incorrect. `errors.Wrap(nil, ...)` is `nil`.
</blockquote></details>

good catch! 

https://github.com/cockroachdb/cockroach/pull/13971

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13908)*
<!-- Sent from Reviewable.io -->
